File: study_play/drills/05_trees_stacks_reflex.js
/**
 * REFLEX DRILL 05 — Trees & Stacks
 *
 * RUN: node study_play/drills/05_trees_stacks_reflex.js
 */

class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

// TODO: REFLEX — inorder traversal (iterative with stack preferred)
function inorderTraversal(root) {
  throw new Error("Implement from memory");
}

// TODO: REFLEX — max depth of binary tree
function maxDepth(root) {
  throw new Error("Implement from memory");
}

// TODO: REFLEX — valid parentheses
function isValidParentheses(s) {
  throw new Error("Implement from memory");
}

// TODO: REFLEX — daily temperatures (next greater to the right) — monotonic stack
function dailyTemperatures(temps) {
  throw new Error("Implement from memory");
}

function assert(name, cond) {
  if (!cond) throw new Error(`FAIL: ${name}`);
  console.log(`PASS: ${name}`);
}

const tree = new TreeNode(1, new TreeNode(2), new TreeNode(3, new TreeNode(4), null));
assert("inorderTraversal", JSON.stringify(inorderTraversal(tree)) === "[2,1,4,3]");
assert("maxDepth", maxDepth(tree) === 3);
assert("isValidParentheses true", isValidParentheses("()[]{}") === true);
assert("isValidParentheses false", isValidParentheses("(]") === false);
assert(
  "dailyTemperatures",
  JSON.stringify(dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73])) ===
    "[1,1,4,2,1,1,0,0]"
);

console.log("\nAll trees/stacks reflex drills passed.");

module.exports = { TreeNode, inorderTraversal, maxDepth, isValidParentheses, dailyTemperatures };

File: study_play/blanks/05_trees_stacks_reflex.go
//go:build ignore

// REFLEX DRILL 05 — Trees & Stacks
//
// RUN: go run ./drills/05_trees_stacks_reflex
package main

import (
	"fmt"
	"reflect"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// TODO: REFLEX — inorder traversal (iterative with stack preferred)
func inorderTraversal(root *TreeNode) []int {
	panic("Implement from memory")
}

// TODO: REFLEX — max depth of binary tree
func maxDepth(root *TreeNode) int {
	panic("Implement from memory")
}

// TODO: REFLEX — valid parentheses
func isValidParentheses(s string) bool {
	panic("Implement from memory")
}

// TODO: REFLEX — daily temperatures (next greater to the right) — monotonic stack
func dailyTemperatures(temps []int) []int {
	panic("Implement from memory")
}

func assert(name string, cond bool) {
	if !cond {
		panic(fmt.Sprintf("FAIL: %s", name))
	}
	fmt.Printf("PASS: %s\n", name)
}

func main() {
	tree := &TreeNode{
		Val:   1,
		Left:  &TreeNode{Val: 2},
		Right: &TreeNode{Val: 3, Left: &TreeNode{Val: 4}},
	}
	assert("inorderTraversal", reflect.DeepEqual(inorderTraversal(tree), []int{2, 1, 4, 3}))
	assert("maxDepth", maxDepth(tree) == 3)
	assert("isValidParentheses true", isValidParentheses("()[]{}") == true)
	assert("isValidParentheses false", isValidParentheses("(]") == false)
	assert("dailyTemperatures", reflect.DeepEqual(
		dailyTemperatures([]int{73, 74, 75, 71, 69, 72, 76, 73}),
		[]int{1, 1, 4, 2, 1, 1, 0, 0},
	))

	fmt.Println("\nAll trees/stacks reflex drills passed.")
}
